src/main/clojure/clara/rules/engine.cljc

```diff
@@ -152,26 +152,34 @@
 (def ^:dynamic *rule-context* nil)
 
 (defn- flush-updates
-  "Flush pending updates in the current session. Returns true if there were some items to flush,
-  false otherwise"
+  "Flush all pending updates in the current session. Returns true if there were
+   some items to flush, false otherwise"
   [current-session]
+  (letfn [(flush-all [current-session flushed-items?]
+            (let [{:keys [rulebase transient-memory transport insertions get-alphas-fn listener]} current-session
+                  pending-updates @(:pending-updates current-session)]
 
-  (let [{:keys [rulebase transient-memory transport insertions get-alphas-fn listener]} current-session
-        pending-updates @(:pending-updates current-session)]
+              ;; Remove the facts here so they are re-inserted if we flush recursively.
+              (reset! (:pending-updates current-session) [])
 
-    ;; Remove the facts here so they are re-inserted if we flush recursively.
-    (reset! (:pending-updates current-session) [])
+              (if (empty? pending-updates)
+                flushed-items?
+                (do
+                  (doseq [partition (partition-by :type pending-updates)
+                          :let [facts (mapcat :facts partition)]
+                          [alpha-roots fact-group] (get-alphas-fn facts)
+                          root alpha-roots]
 
-    (doseq [partition (partition-by :type pending-updates)
-            :let [facts (mapcat :facts partition)]
-            [alpha-roots fact-group] (get-alphas-fn facts)
-            root alpha-roots]
+                    (if (= :insert (:type (first partition)))
+                      (alpha-activate root fact-group transient-memory transport listener)
+                      (alpha-retract root fact-group transient-memory transport listener)))
 
-      (if (= :insert (:type (first partition)))
-        (alpha-activate root fact-group transient-memory transport listener)
-        (alpha-retract root fact-group transient-memory transport listener)))
+                  ;; There may be new :pending-updates due to the flush just
+                  ;; made.  So keep flushing until there are none left.  Items
+                  ;; were flushed though, so flush-items? is now true.
+                  (flush-all current-session true)))))]
```

Comment:
We talked about this some in person, but basically it is not a very realistic scenario to get into a situation where this `flush-all` recursion would go beyond a small number of iterations.  Due to this, not fixing it right now.  The only new `:pending-updates` that can occur on each iteration is new retractions being added.  The way to get this flushing cycle to go on a long time would be to have many "chained" logical dependencies among facts.
Example:
```clj
(let [n 10
      ss (mapv str (range n))
      rhss (mapv (fn [s] `(insert! ~s))
                 ss)
      rules (mapv (fn [s rhs]
                    {:lhs (if (= ::none s)
                            []
                            [{:type `String
                              :constraints [`(= ~'this ~s)]}])
                     :rhs rhs})
                  (cons ::none ss)
                  rhss)
      q (dsl/parse-query [] [[?ss <- (acc/all) :from [String]]])
      s (mk-session rules [q])
      before (-> s
                 fire-rules)
      after (-> before
                (retract "0")
                fire-rules)]
  [(-> (query before q) first :?ss count)
   (-> (query after q) first :?ss count)])

;;= [10 0]

```

Changing `n` to 1000 here, does indeed give a StackOverflowError however, it is not due to this change (happens in master and 0.11.1 etc).  It is due to lazy collections being used in a few places of the engine.

Examples being:
1. https://github.com/rbrush/clara-rules/blob/0.11.1/src/main/clojure/clara/rules/engine.cljc#L343
2. https://github.com/rbrush/clara-rules/blob/0.11.1/src/main/clojure/clara/rules/engine.cljc#L390
3. https://github.com/rbrush/clara-rules/blob/0.11.1/src/main/clojure/clara/rules/engine.cljc#L271

1 and 2 are very common and making them eager could have adverse side effects to performance - so that would warrant its own issue.  3 is a bit easier, but again it'd be worthwhile to assess where laziness is helping and where it is not.

Also, this rule situation I just described is not something I'd expect to see in practice.  It would not be likely that rules structured in this way would be very coherent or even be able to perform very well with such a deeply linked dependency chain among the rules.

So for now, keeping this `flush-all` as is for now unless there are objections.  The lazy collections comes up before we even get to this flushing loop having issues.  Changing to loop-recur here would be fairly trivial at a later time.